### ISSUE TITLE
Using a keybind to open WCT whilst it's in another mode bugs out the terminal, inserting items into it in this state causes all terminals on the server to become blank

### ISSUE THREAD

**Issue by Dazzuh** [p455w0rd/AE2WirelessTerminalLibrary#16]
This is sorta 2 issues rolled into one. 

Issue one being that when you use a keybind to open wireless crafting terminal whilst the terminal was in another mode by either shift-rightclicking in your hand or using the switch terminal button in the GUI, it opens with a crafting terminal interface but it doesn't seem to actually switch to crafting terminal mode, resulting in crafting terminal GUI but the button still displays the option to swap to WCT: 
![image](https://user-images.githubusercontent.com/3359779/93820310-dfc67880-fc54-11ea-9c88-4abd4351d3a5.png)

Meaning, in the above situation, the terminal still seems to think it's in fluid mode. 

This leads to issue two, when you insert items into the terminal whilst it's in this state (specifically with fluid mode) it causes _all_ terminals on the server (including normal AE2 terminals) to become blank, not showing any items until the server is restarted. 

This issue was discussed a few months ago in an AE2 issue, not sure why it was never posted here too: https://github.com/AppliedEnergistics/Applied-Energistics-2/issues/4380

**Comment by malcom2073** [comment]
Relevant stack trace from the server @Dazzuh was on, happened I believe just when they tried to insert items into the wireless terminal, after which no terminals (wired or wireless) worked.

```
java.util.concurrent.ExecutionException: java.lang.ClassCastException: appeng.util.item.AEItemStack cannot be cast to appeng.api.storage.data.IAEFluidStack
	at java.util.concurrent.FutureTask.report(FutureTask.java:122) ~[?:1.8.0_265]
	at java.util.concurrent.FutureTask.get(FutureTask.java:192) ~[?:1.8.0_265]
	at net.minecraft.util.Util.func_181617_a(SourceFile:47) [h.class:?]
	at net.minecraft.server.MinecraftServer.func_71190_q(MinecraftServer.java:723) [MinecraftServer.class:?]
	at net.minecraft.server.dedicated.DedicatedServer.func_71190_q(DedicatedServer.java:397) [nz.class:?]
	at net.minecraft.server.MinecraftServer.func_71217_p(MinecraftServer.java:668) [MinecraftServer.class:?]
	at net.minecraft.server.MinecraftServer.run(MinecraftServer.java:526) [MinecraftServer.class:?]
	at java.lang.Thread.run(Thread.java:748) [?:1.8.0_265]
Caused by: java.lang.ClassCastException: appeng.util.item.AEItemStack cannot be cast to appeng.api.storage.data.IAEFluidStack
	at appeng.fluids.util.FluidList.findPrecise(FluidList.java:33) ~[FluidList.class:?]
	at appeng.me.storage.BasicCellInventory.extractItems(BasicCellInventory.java:221) ~[BasicCellInventory.class:?]
	at appeng.me.storage.MEPassThrough.extractItems(MEPassThrough.java:63) ~[MEPassThrough.class:?]
	at appeng.me.storage.MEInventoryHandler.extractItems(MEInventoryHandler.java:117) ~[MEInventoryHandler.class:?]
	at appeng.me.storage.MEInventoryHandler.extractItems(MEInventoryHandler.java:117) ~[MEInventoryHandler.class:?]
	at appeng.me.storage.DriveWatcher.extractItems(DriveWatcher.java:77) ~[DriveWatcher.class:?]
	at appeng.me.storage.NetworkInventoryHandler.injectItems(NetworkInventoryHandler.java:99) ~[NetworkInventoryHandler.class:?]
	at appeng.me.cache.NetworkMonitor.injectItems(NetworkMonitor.java:157) ~[NetworkMonitor.class:?]
	at p455w0rd.ae2wtlib.helpers.WTGuiObjectImpl.injectItems(WTGuiObjectImpl.java:203) ~[WTGuiObjectImpl.class:?]
	at appeng.util.Platform.poweredInsert(Platform.java:1229) ~[Platform.class:?]
	at appeng.util.Platform.poweredInsert(Platform.java:1218) ~[Platform.class:?]
	at p455w0rd.wct.container.ContainerWCT.shiftStoreItem(ContainerWCT.java:1143) ~[ContainerWCT.class:?]
	at p455w0rd.wct.container.ContainerWCT.func_82846_b(ContainerWCT.java:989) ~[ContainerWCT.class:?]
	at net.minecraft.inventory.Container.func_184996_a(Container.java:246) ~[afr.class:?]
	at p455w0rd.wct.container.ContainerWCT.func_184996_a(ContainerWCT.java:1154) ~[ContainerWCT.class:?]
	at net.minecraft.network.NetHandlerPlayServer.func_147351_a(NetHandlerPlayServer.java:1162) ~[pa.class:?]
	at net.minecraft.network.play.client.CPacketClickWindow.func_148833_a(CPacketClickWindow.java:38) ~[lf.class:?]
	at net.minecraft.network.play.client.CPacketClickWindow.func_148833_a(CPacketClickWindow.java:12) ~[lf.class:?]
	at net.minecraft.network.PacketThreadUtil$1.run(SourceFile:13) ~[hv$1.class:?]
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) ~[?:1.8.0_265]
	at java.util.concurrent.FutureTask.run(FutureTask.java:266) ~[?:1.8.0_265]
	at net.minecraft.util.Util.func_181617_a(SourceFile:46) ~[h.class:?]
	... 5 more

```

**Comment by ChromaPIE** [comment]
It's such an annoying bug. And it's disappointing to see this project being out of maintenance. Thanks for letting us know what is the trigger of the bug.

**Comment by machineonamission** [comment]
My friend accidentally reproduced it on our All The Mods 3:Remix server. Glad to know the trigger so i can avoid doing that.

**Comment by MrYang-Yang** [comment]
Is there a solution to this serious problem?

**Comment by ruifung** [comment]
Did some digging:

1. Opening with keybinds are handled by the individual Wireless terminal mods, those mods straight open the keybinds with the WirelessUltimateTerminal item.
2. WirelessUltimateTerminal item implementation for getStorageChannel() depends on NBT data on the WUT ItemStack and thus if it is in a different mode, it will return the StorageChannel for the mode defined in its NBT.
3. Opening via keybind thus results in the GUI container being opened with potentially the wrong storage channel.

Basically: Don't put the fluid terminal into your ultimate terminal.
